layer_analyzer.go

```diff
@@ -0,0 +1,124 @@
+package lifecycle
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/buildpack"
+	"github.com/buildpacks/lifecycle/cmd"
+	"github.com/buildpacks/lifecycle/launch"
+	"github.com/buildpacks/lifecycle/platform"
+)
+
+type LayerAnalyzer interface {
+	Analyze(buildpacks []buildpack.GroupBuildpack, skipLayers bool, appMeta platform.LayersMetadata, cache Cache) (platform.CacheMetadata, error)
+}
+
+type DefaultLayerAnalyzer struct {
+	Logger            Logger
+	LayersDir         string
+	MetadataRetriever CacheMetadataRetriever
+	Platform          cmd.Platform
+}
+
+func NewLayerAnalyzer(logger Logger, metadataRetriever CacheMetadataRetriever, layersDir string, platform cmd.Platform) LayerAnalyzer {
+	return &DefaultLayerAnalyzer{
+		LayersDir:         layersDir,
+		Logger:            logger,
+		MetadataRetriever: metadataRetriever,
+		Platform:          platform,
+	}
+}
+
+func (la *DefaultLayerAnalyzer) Analyze(buildpacks []buildpack.GroupBuildpack, skipLayers bool, appMeta platform.LayersMetadata, cache Cache) (platform.CacheMetadata, error) {
```

Comment:
Since this function is providing both `appMeta` and `cacheMeta` I think we should either pass a pointer to both variables as an argument to the function, or return 3 values. It's a bit weird that one is passed in and the other is returned.

Reply:
This function isn't providing `appMeta`, it is consumes it.

Reply:
Ohh sorry, I was confusing this with `DecodeLabel` :/

Reply:
Thinking about this some more... what about moving `CacheMetadataRetriever` from `LayerAnalyzer` to `Analyzer`, so that it behaves more like `Restorer` (see [here](https://github.com/buildpacks/lifecycle/pull/582/files#diff-98f352cc6ef347aa505fe6d5f3e01c3a47ee0b5d1a91d1e918343284c893b51eR28)). Then `appMeta` and `cacheMeta` could both be passed in, and we eliminate some of the weirdness around ignoring the return value from `LayerAnalyzer.Analyze()`.

Reply:
I actually did have it like that in the past but I didn't like juggling around 2 fields `LayerAnalyzer` and `CacheMetadataRetriever` that were both > 0.7. AND you needed to know to use one before the other. I like hiding the `CacheMetadataRetriever` behind the a purposeful interface like `Analyze`. It also makes it a lot easier to mock when testing.

Reply:
Hmm... the thing is, `CacheMetadataRetriever` is really just a fancy wrapper around `cache.RetrieveMetadata()`. Maybe we just wrap it in a function, so that we don't have to pass a struct around? IDK.

Reply:
I don't feel super strongly about this, just throwing out ideas...

Reply:
I feel like that might be worth doing as part of this PR. I recall the original PR had that. It would reduce some of the duplication that we are now creating with the tests...

Reply:
I pushed up the `LayerMetadataRestorer` tests from the other PR

Reply:
I put up a PR to this PR (https://github.com/buildpacks/lifecycle/pull/583) that moves the cache metadata retrieval around. I don't **think** it would make our life harder for mocking out the layer metadata restorer, because the cache metadata is passed in. It wouldn't make the cache any easier to mock in analyzer or restorer tests, but since we are already NOT mocking the cache I feel okay about it. Let me know what you think!